### ISSUE TITLE
Issue certificate from the web

### ISSUE THREAD

**Issue by redNixon** [mathiasertl/django-ca#22]
In the documentation, this is listed as a description of the web interface:

    The web interface allows you to perform the most common tasks necessary when running certificate authority. It is implemented using Djangos admin interface. You can:

    Issue and revoke certificates.
    ...

However, I do not see a way to create new certificates from the webUI.  Has this functionality not been implemented yet or am I just using/reading this wrong?

**Comment by redNixon** [comment]
...and after looking at this for far too many hours, a valid rootCA was all that was needed to make the CSR signing page show up.  I was attempting to use a pre-existing certificate instead of using init_ca, incorrectly it appears.

This is very much a user error on my part, however a warning on the certificate page about a misconfigured site would be helpful for those trying to quickly demo using a pre-existing certificate.

**Comment by mathiasertl** [comment]
A warning sounds like a very good idea, but I need a little more information. Not having a local root CA at all is deliberately not a problem for the web interface. We have a setup were a dedicated machine has the root CA, and the web interface is on a different machine only used for revoking certificates, configuring notification emails and OCSP.

But you say "was attempting to use a pre-existing certificate instead of using init_ca" which seems to imply that you did configure a root CA? Can you tell me more about the setup you were attempting so I can catch the misconfiguration more precisely?

This reminds me that I need to add a feature to import an existing CA. The difference is that it's public key and fingerprint is stored in the database (also, the signed Certificates have a Foreign Key to it).

**Comment by mathiasertl** [comment]
Btw, if need to import a CA right now, you should be able to do so by starting a python shell with `manage.py shell` (tipp: install ipython before doing so!) and doing:

```
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from django_ca.models import CertificateAuthority

with open('/path/to/public_key.pem', 'rb') as stream:
    pem = x509.load_pem_x509_certificate(pem, default_backend()
ca = CertificateAuthority(name="name you can change", private_key_path='path/to/private_key')
ca.x509 = pem
ca.save()
ca.serial  # should output the serial for your CA
```

If you have an intermediate CA, pass the parent to the constructor:

```
parent = CertificateAuthority(....)
child = CertificateAuthority(..., parent=parent)
```

**Comment by redNixon** [comment]
Correct, I already have a pre-existing root CA that I am trying to use with django-ca.  

The "quick and dirty" method I thought of to get it imported was to run init_ca with the default values then replace the values it created in the DB with my own after the fact.  *Something* with wrong in my hackjob which caused it to think that there was no root CA at all which caused the "Add" button to not appear.  Attempts to figure out what was going on were compounded by having the Jet admin interface app installed which breaks the certificate/add form.  Starting the project over with the default django settings allowed it to work properly.

When you are a first time user and the app is in this state it is hard to figure what exactly it does or is suppose to do.  My usecase/misconfiguration is probably not common, but running migrate && runserver is probably a common  demo techique which would get you into the same state.  I could see the code to sign certificates, but I could not follow a clear path from the webui code to the signer.

Based on your comment and the issue I ran into when setting it up for the first time, these are the two basic requirements I gathered:
1:  Allow the ability to distribute the signing and management functionality without having superfluous elements on the UI that do not add any functionality.
2 :  Give visual feedback when the user expects to see UI controls but due to some reason(misconfiguration, incomplete setup, expired CA, etc), they are not usable.

I believe the workflow most users that go from demo to production will start off with a single server for basic validation checks then move to setting up the secure infrastructure and segregation needed to move it into production(I’m just going to hope everyone does not have their signing keys sitting on a public interface).

I think easiest way to support the requirements and the workflow would be to modularize the functionality(OCSP, watchers, certificate revoke, cert renewal, cert creation, etc) and use the module setting( ie “Certificate Creation” ) as the test of whether or not to show the UI controls and enable the controller functionality.  A misconfigured site will still not work, but if a user goes to site.com/admin/certificates/add and sees that there is no root certificate or a warning besides it saying that it is expired then the reason for the failure and the needed remediation would be fairly intuitive.

 I am already changing the certificate page to make it easier/more intuitive to use in order to fit my usecase(multiuser env, userbase with low/no cert knowledge).  I am also going to be adding some other modules to django-ca to expand it from a pure x509 CA as well(openSSH signed keys do not use true x509 they also handle revocation differently).  I can put in some pull requests the UI to make the generic usecase more intuitive as well as adding some modularization.  Those two changes should probably be tracked in separate issues assuming you are ok with the general premise.
